[JSC] Optimize searching untyped element from int32 array for `indexOf` and `includes`

sosukesuzuki opened the pull request:
#### 3e9331148002ecc61cc0a11e2084e58689d5f5a2
<pre>
[JSC] Optimize searching untyped element from int32 array for `indexOf` and `includes`
<a href="https://bugs.webkit.org/show_bug.cgi?id=290388">https://bugs.webkit.org/show_bug.cgi?id=290388</a>

Reviewed by Yusuke Suzuki.

This patch adds a dedicated JIT operation for searching untyped
element from int32 array for `Array#indexOf` and `Array#includes`.

                                                TipOfTree                  Patched

array-prototype-indexOf-untyped-int32        36.1595+-0.6719     ^      6.9764+-0.0820        ^ definitely 5.1831x faster
array-prototype-includes-untyped-int32       39.2141+-0.4759     ^      8.3772+-0.0831        ^ definitely 4.6811x faster

* JSTests/microbenchmarks/array-prototype-includes-untyped-int32.js: Added.
(shouldBe):
(test):
* JSTests/microbenchmarks/array-prototype-indexOf-untyped-int32.js: Added.
(shouldBe):
(test):
* JSTests/stress/array-prototype-includes-untyped-int32-hole.js: Added.
(shouldBe):
(test):
* JSTests/stress/array-prototype-indexOf-untyped-int32-hole.js: Added.
(shouldBe):
(test):
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileArrayIndexOfOrArrayIncludes):

Canonical link: <a href="https://commits.webkit.org/292999@main">https://commits.webkit.org/292999@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a87f35017c633504515bf0e312dc6e099bbffb8e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97673 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17298 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7515 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102779 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48202 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17592 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25751 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74390 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31585 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100676 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13328 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/88305 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54739 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13097 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6211 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47644 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/90349 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/83113 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6288 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104781 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/96296 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24753 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/18049 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/83444 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25125 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84455 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82875 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20869 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27441 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/5124 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/18351 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24714 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/29883 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/119921 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24536 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/33661 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27850 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26110 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/dfg/DFGOperations.cpp:
```diff
@@ -3942,6 +3942,37 @@ JSC_DEFINE_JIT_OPERATION(operationArrayIncludesValueInt32OrContiguous, UCPUStric
+        if (!value || !value.isNumber())
+            continue;
+        if (searchElement.asInt32() == value.asNumber())
+            OPERATION_RETURN(scope, toUCPUStrictInt32(1));
```
Ditto.

sosukesuzuki replied:
Fixed!

Constellation commented on Source/JavaScriptCore/dfg/DFGOperations.cpp:
```diff
@@ -4096,6 +4127,35 @@ JSC_DEFINE_JIT_OPERATION(operationArrayIndexOfValueInt32OrContiguous, UCPUStrict
+        JSValue value = data[index].get();
+        if (!value || !value.isNumber())
+            continue;
+        if (searchElement.asInt32() == value.asNumber())
```
`searchElement` is not guaranteed as `Int32`. So you cannot use `asInt32`.
You should check whether the contained content is Int32 via `isInt32AsAnyInt` and `asInt32AsAnyInt`.
On the other hand, value is ensured as int32, so you should use `asInt32()`.

sosukesuzuki replied:
Fixed!

Constellation commented on Source/JavaScriptCore/dfg/DFGOperations.cpp:
```diff
@@ -4096,6 +4127,35 @@ JSC_DEFINE_JIT_OPERATION(operationArrayIndexOfValueInt32OrContiguous, UCPUStrict
+
+    for (; index < length; ++index) {
+        JSValue value = data[index].get();
+        if (!value || !value.isNumber())
```
If Array is Int32, then value is always int32, so no need to use `isNumber()` check.

sosukesuzuki replied:
Fxied!

Constellation commented on Source/JavaScriptCore/dfg/DFGOperations.cpp:
```diff
@@ -3942,6 +3942,37 @@ JSC_DEFINE_JIT_OPERATION(operationArrayIncludesValueInt32OrContiguous, UCPUStric
+
+    for (; index < length; ++index) {
+        JSValue value = data[index].get();
+        if (!value || !value.isNumber())
```
Ditto.

sosukesuzuki replied:
Fixed!

Constellation commented on Source/JavaScriptCore/dfg/DFGOperations.cpp:
```diff
@@ -3942,6 +3942,37 @@ JSC_DEFINE_JIT_OPERATION(operationArrayIncludesValueInt32OrContiguous, UCPUStric
+        OPERATION_RETURN(scope, toUCPUStrictInt32(1));
+
+    if (!searchElement.isNumber())
+        OPERATION_RETURN(scope, toUCPUStrictInt32(0));
```
Ditto.

sosukesuzuki replied:
Fixed!

Constellation commented on Source/JavaScriptCore/dfg/DFGOperations.cpp:
```diff
@@ -3942,6 +3942,39 @@ JSC_DEFINE_JIT_OPERATION(operationArrayIncludesValueInt32OrContiguous, UCPUStric
+
+    int32_t searchInt32 = searchElement.asInt32AsAnyInt();
+
+    for (; index < length; ++index) {
```
I wonder if we should use find64 with JSValue::encode(intValue).

darinadler replied:
Seems to me that we definitely should!

Constellation commented on Source/JavaScriptCore/dfg/DFGOperations.cpp:
```diff
@@ -4096,6 +4129,37 @@ JSC_DEFINE_JIT_OPERATION(operationArrayIndexOfValueInt32OrContiguous, UCPUStrict
+
+    int32_t searchInt32 = searchElement.asInt32AsAnyInt();
+
+    for (; index < length; ++index) {
```
Ditto

Constellation commented on Source/JavaScriptCore/dfg/DFGOperations.cpp:
```diff
@@ -3942,6 +3942,45 @@ JSC_DEFINE_JIT_OPERATION(operationArrayIncludesValueInt32OrContiguous, UCPUStric
+    }
+
+    int32_t searchInt32 = searchElement.asInt32AsAnyInt();
+    for (; index < length; ++index) {
```
I think what we should do here is always using `WTF::find64`. Let's do,

```
EncodedJSValue encodedSearchElement = JSValue::encode(jsNumber(searchElement.asInt32AsAnyInt()));
```
And use `encodedSearchElement` with `WTF::find64`, so we do not need this loop twice.

sosukesuzuki replied:
Looks great! Fixed

sosukesuzuki committed:
[JSC] Optimize searching untyped element from int32 array for `indexOf` and `includes`
https://bugs.webkit.org/show_bug.cgi?id=290388

Reviewed by Yusuke Suzuki.

This patch adds a dedicated JIT operation for searching untyped
element from int32 array for `Array#indexOf` and `Array#includes`.

                                                TipOfTree                  Patched

array-prototype-indexOf-untyped-int32        36.1595+-0.6719     ^      6.9764+-0.0820        ^ definitely 5.1831x faster
array-prototype-includes-untyped-int32       39.2141+-0.4759     ^      8.3772+-0.0831        ^ definitely 4.6811x faster

* JSTests/microbenchmarks/array-prototype-includes-untyped-int32.js: Added.
(shouldBe):
(test):
* JSTests/microbenchmarks/array-prototype-indexOf-untyped-int32.js: Added.
(shouldBe):
(test):
* JSTests/stress/array-prototype-includes-untyped-int32-hole.js: Added.
(shouldBe):
(test):
* JSTests/stress/array-prototype-indexOf-untyped-int32-hole.js: Added.
(shouldBe):
(test):
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileArrayIndexOfOrArrayIncludes):

Canonical link: https://commits.webkit.org/292999@main
